File: internal/router/team.go
package router

import (
	"github.com/elabosak233/cloudsdale/internal/controller"
	"github.com/elabosak233/cloudsdale/internal/model"
	"github.com/elabosak233/cloudsdale/internal/service"
	"github.com/elabosak233/cloudsdale/internal/utils/convertor"
	"github.com/gin-gonic/gin"
	"net/http"
)

type ITeamRouter interface {
	Register()
}

type TeamRouter struct {
	router     *gin.RouterGroup
	controller controller.ITeamController
}

func NewTeamRouter(teamRouter *gin.RouterGroup, teamController controller.ITeamController) ITeamRouter {
	return &TeamRouter{
		router:     teamRouter,
		controller: teamController,
	}
}

func (t *TeamRouter) Register() {
	t.router.GET("/", t.controller.Find)
	t.router.POST("/", t.controller.Create)
	t.router.DELETE("/:id", t.CanModifyTeam(), t.controller.Delete)
	t.router.PUT("/:id", t.CanModifyTeam(), t.controller.Update)
	t.router.POST("/:id/users", t.controller.CreateUser)
	t.router.DELETE("/:id/users/:user_id", t.CanModifyTeam(), t.controller.DeleteUser)
	t.router.GET("/:id/invite", t.CanModifyTeam(), t.controller.GetInviteToken)
	t.router.PUT("/:id/invite", t.CanModifyTeam(), t.controller.UpdateInviteToken)
	t.router.POST("/:id/join", t.controller.Join)
	t.router.POST("/:id/leave", t.controller.Leave)
	t.router.POST("/:id/avatar", t.CanModifyTeam(), t.controller.SaveAvatar)
	t.router.DELETE("/:id/avatar", t.CanModifyTeam(), t.controller.DeleteAvatar)
}

func (t *TeamRouter) CanModifyTeam() gin.HandlerFunc {
	return func(ctx *gin.Context) {
		user := ctx.MustGet("user").(*model.User)

		if ok := service.S().AuthService.CanModifyTeam(user, convertor.ToUintD(ctx.Param("id"), 0)); !ok {
			ctx.JSON(http.StatusForbidden, gin.H{
				"code": http.StatusForbidden,
			})
			ctx.Abort()
		}

		ctx.Next()
	}
}

File: src/server/router/team.rs
use axum::{
    extract::DefaultBodyLimit,
    middleware::from_fn,
    routing::{delete, get, post, put},
    Router,
};

use crate::server::{controller, middleware::auth};
use crate::util::jwt::Group;

pub fn router() -> Router {
    return Router::new()
        .route("/", get(controller::team::find).layer(from_fn(auth::jwt(Group::Guest))))
        .route("/", post(controller::team::create).layer(from_fn(auth::jwt(Group::User))))
        .route("/:id", put(controller::team::update).layer(from_fn(auth::jwt(Group::User))))
        .route("/:id", delete(controller::team::delete).layer(from_fn(auth::jwt(Group::User))))
        .route("/:id/users", post(controller::team::create_user).layer(from_fn(auth::jwt(Group::User))))
        .route(
            "/:id/users/:user_id",
            delete(controller::team::delete_user).layer(from_fn(auth::jwt(Group::User))),
        )
        .route("/:id/invite", get(controller::team::get_invite_token).layer(from_fn(auth::jwt(Group::User))))
        .route("/:id/invite", put(controller::team::update_invite_token).layer(from_fn(auth::jwt(Group::User))))
        .route("/:id/join", post(controller::team::join).layer(from_fn(auth::jwt(Group::User))))
        .route("/:id/leave", delete(controller::team::leave).layer(from_fn(auth::jwt(Group::User))))
        .route("/:id/avatar", get(controller::team::find_avatar))
        .route("/:id/avatar/metadata", get(controller::team::find_avatar_metadata))
        .route(
            "/:id/avatar",
            post(controller::team::save_avatar)
                .layer(DefaultBodyLimit::max(3 * 1024 * 1024 /* MB */))
                .layer(from_fn(auth::jwt(Group::User))),
        )
        .route("/:id/avatar", delete(controller::team::delete_avatar).layer(from_fn(auth::jwt(Group::User))));
}
